ffi/src/handle.rs

```diff
@@ -1,235 +1,616 @@
-/// }
-///
-/// // Connect the two
-/// impl SizedArcHandle for MyStructHandle {
-///     type Target = MyStruct;
-/// }
-///
-/// fn sized_handle_example(val: Arc<MyStruct>) {
-///     let handle: *const MyStructHandle = ArcHandle::into_handle(val);
-///     let val2 = unsafe { ArcHandle::clone_as_arc(handle) };
-///     unsafe { ArcHandle::drop_handle(handle) };
-///     // `handle` is no longer safe to use, but `val2` (a cloned Arc) is still valid
-/// }
-/// ```
-///
-/// # Safety
-///
-/// In addition to being a raw pointer, a handle always points to an underlying allocation of some
-/// other type. Thus, it is _ALWAYS_ incorrect to dereference (`*`) a handle or cast it to any other
-/// type, (including `Target`). The only safe operations on a handle are to copy the pointer itself
-/// by value and to invoke associated functions of `ArcHandle`.
-///
-/// Additionally, this trait intentionally does _NOT_ expose anything like an `as_ref()` function,
-/// because the resulting reference would have an arbitrary (= unnecessarily dangerous) lifetime
-/// that cannot be safely enforced, even if we aggressively bound its lifetime. For example:
-/// ```
-/// let handle: *const MyHandle = ArcHandle::into_handle(...);
-/// let handle_copy = handle; // does NOT increment the Arc refcount
-/// let my_ref = ArcHandle::as_ref(&handle);
-/// ArcHandle::drop_handle(handle_copy); // still-borrowed `handle` can't prevent this
-/// my_ref.some_access(); // Illegal access to dangling reference
-/// ```
-///
-/// If a borrowed reference is needed, call [ArcHandle::clone_as_arc] and invoke [Arc::as_ref] on
-/// the result.
-/// ```
-/// let handle: *const MyHandle = ArcHandle::into_handle(...);
-/// let handle_copy = handle; // does NOT increment the Arc refcount
-/// let my_ref = ArcHandle::clone_as_arc(handle); // increments refcount
-/// ArcHandle::drop_handle(handle);
-/// my_ref.some_access(); // No problem.
-/// ```
-///
-/// # Synchronization
-///
-/// The handle (a raw pointer) is neither [Send] nor [Sync] by default:
-///
-/// It is ONLY safe to manually implement [Send] for a handle (or a type that embeds it) if all of
-/// the following conditions hold:
-///
-/// * The handle's `Target` type is `Send`
-///
-/// * Calls to [ArcHandle::clone_as_arc] and [ArcHandle::drop_handle] always obey their respective
-///   safety requirements. In particular, this means proving (or at least affirming) that no handle
-///   is ever accessed after passing it to [ArcHandle::drop_handle].
-///
-/// It is ONLY safe to implement [Sync] for a handle (or type embedding it) if all of the following
-/// conditions hold:
-///
-/// * The `Target` type is `Sync`
-///
-/// * The handle is (correctly) `Send`
-pub trait ArcHandle: Sized {
-    /// The target type this handle represents.
-    type Target: ?Sized + Send + Sync;
+pub trait HandleDescriptor {
+    /// The true type of the handle's underlying object
+    type Target: ?Sized + Send;
```

Comment:
NOTE: Mutable handles only need to be `Send` (not `Sync`), because any use by external code that would rely on them being `Sync` would actually break the `Send` contract: Mutable handles are mutable references, and because handles are opaque to extern code, rust FFI code always receives the handle as mutable. Doing that from multiple threads means multiple mutable reference now exist in rust at the same time, which produces undefined behavior.